lib/serializer/BinaryDeserializer.h

```diff
@@ -0,0 +1,535 @@
+			});
+		}
+	};
+
+	template<typename Ser,typename T>
+	struct LoadIfStackInstance
+	{
+		static bool invoke(Ser &s, T &data)
+		{
+			return false;
+		}
+	};
+
+	template<typename Ser>
+	struct LoadIfStackInstance<Ser, CStackInstance *>
+	{
+		static bool invoke(Ser &s, CStackInstance* &data)
+		{
+			CArmedInstance *armedObj;
+			SlotID slot;
+			s.load(armedObj);
+			s.load(slot);
+			if(slot != SlotID::COMMANDER_SLOT_PLACEHOLDER)
+			{
+				assert(armedObj->hasStackAtSlot(slot));
+				data = armedObj->stacks[slot];
+			}
+			else
+			{
+				auto hero = dynamic_cast<CGHeroInstance *>(armedObj);
+				assert(hero);
+				assert(hero->commander);
+				data = hero->commander;
+			}
+			return true;
+		}
+	};
+
+	template <typename T, typename Enable = void>
+	struct ClassObjectCreator
+	{
+		static T *invoke()
+		{
+			static_assert(!std::is_abstract<T>::value, "Cannot call new upon abstract classes!");
+			return new T();
+		}
+	};
+
+	template<typename T>
+	struct ClassObjectCreator<T, typename std::enable_if<std::is_abstract<T>::value>::type>
+	{
+		static T *invoke()
+		{
+			throw std::runtime_error("Something went really wrong during deserialization. Attempted creating an object of an abstract class " + std::string(typeid(T).name()));
+		}
+	};
+
+#define READ_CHECK_U32(x)			\
+	ui32 length;			\
+	load(length);				\
+	if(length > 500000)				\
+	{								\
+		logGlobal->warnStream() << "Warning: very big length: " << length;\
+		reader->reportState(logGlobal);			\
+	};
+
+	template <typename T> class CPointerLoader;
+
+	class CBasicPointerLoader
+	{
+	public:
+		virtual const std::type_info * loadPtr(CLoaderBase &ar, void *data, ui32 pid) const =0; //data is pointer to the ACTUAL POINTER
+		virtual ~CBasicPointerLoader(){}
+
+		template<typename T> static CBasicPointerLoader *getApplier(const T * t=nullptr)
+		{
+			return new CPointerLoader<T>();
+		}
+	};
+
+	template <typename T> class CPointerLoader : public CBasicPointerLoader
+	{
+	public:
+		const std::type_info * loadPtr(CLoaderBase &ar, void *data, ui32 pid) const override //data is pointer to the ACTUAL POINTER
+		{
+			BinaryDeserializer &s = static_cast<BinaryDeserializer&>(ar);
+			T *&ptr = *static_cast<T**>(data);
+
+			//create new object under pointer
+			typedef typename std::remove_pointer<T>::type npT;
+			ptr = ClassObjectCreator<npT>::invoke(); //does new npT or throws for abstract classes
+			s.ptrAllocated(ptr, pid);
+			//T is most derived known type, it's time to call actual serialize
+			assert(s.fileVersion != 0);
+			ptr->serialize(s,s.fileVersion);
+			return &typeid(T);
+		}
+	};
+
+	CApplier<CBasicPointerLoader> applier;
+
+	int write(const void * data, unsigned size);
+
+public:
+	bool reverseEndianess; //if source has different endianness than us, we reverse bytes
+	si32 fileVersion;
+
+	std::map<ui32, void*> loadedPointers;
+	std::map<ui32, const std::type_info*> loadedPointersTypes;
+	std::map<const void*, boost::any> loadedSharedPointers;
```

Comment:
Details? This map is used to find already created shared_ptr for newly loaded pointer. I don't see how it may cause memory leak. If there is an element in this map then there is a loaded shared_ptr with same pointer.